### DISCUSSION TITLE
Plugins in Official Tiddlywiki.com

### DISCUSSION THREAD

**Discussion by kookma** [TiddlyWiki/TiddlyWiki5#6121] [Q&A]
The official https://tiddlywiki.com/ has few plugins as below! This is very important to ship it with minimum required plugins, but my question is: why Evernote is included? and why not highlight.js is include? Is there any reason?

By the way I appreciate this cleanup! Thank you Jeremy!

![image](https://user-images.githubusercontent.com/830394/137448151-708ef9f6-2b6e-4f20-9e1d-4746d44c3a04.png)

**Comment by nilslindemann** [comment]
Probably they didn't include the highlight.js plugin, because there are not many non-WikiText code examples in the docs and the current highlight.js library is sized 111 kb. If i am not wrong, there is currently no highlighter for WikiText, which would be the main use case.

BTW, I fixed some crude display errors in my local TiddlyWiki by
* updating [highlight.pack.js](https://github.com/Jermolene/TiddlyWiki5/blob/122de63b71a2454314365ac396e051ad1db77641/plugins/tiddlywiki/highlight/files/highlight.pack.js) with the [current version](https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.3.1/highlight.min.js).
* replacing [this line](https://github.com/Jermolene/TiddlyWiki5/blob/122de63b71a2454314365ac396e051ad1db77641/plugins/tiddlywiki/highlight/highlightblock.js#L33) with `hljs.highlightElement(domNode);` and
* [this line](https://github.com/Jermolene/TiddlyWiki5/blob/122de63b71a2454314365ac396e051ad1db77641/plugins/tiddlywiki/highlight/highlightblock.js#L36) with `domNode.children[0].innerHTML = hljs.highlight(text, {language: language}).value;`.

But I didn't test it on the current Wiki, with its p-soup :-p

**Reply by nilslindemann** [reply]
Regarding the Evernote plugin, I guess this is to let Evernote users test directly in the browser how it looks when they import their Evernote data into TiddlyWiki.

**Comment by kookma** [comment]
> BTW, I fixed some crude display errors in my local TiddlyWiki by
> 
> * updating [highlight.pack.js](https://github.com/Jermolene/TiddlyWiki5/blob/122de63b71a2454314365ac396e051ad1db77641/plugins/tiddlywiki/highlight/files/highlight.pack.js) with the [current version](https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.3.1/highlight.min.js).
> * replacing [this line](https://github.com/Jermolene/TiddlyWiki5/blob/122de63b71a2454314365ac396e051ad1db77641/plugins/tiddlywiki/highlight/highlightblock.js#L33) with `hljs.highlightElement(domNode);` and
> * [this line](https://github.com/Jermolene/TiddlyWiki5/blob/122de63b71a2454314365ac396e051ad1db77641/plugins/tiddlywiki/highlight/highlightblock.js#L36) with `domNode.children[0].innerHTML = hljs.highlight(text, {language: language}).value;`.

Great work! I am not very familiar with JS and do not understand your modification, but I appreciate you efforts!

Please have a look at https://github.com/Jermolene/TiddlyWiki5/pull/4502

I am sure @Jermolene welcome to update highlight.js! For me it is important as I use Fortran syntax highlights and the current TW version of Hghlight.js has bug!

Please also have a look at https://github.com/Jermolene/TiddlyWiki5/issues/5961
I am not sure how much v11 changed from v9 as @pmario stated in #5961

**Comment by pmario** [comment]
>I am not sure how much v11 changed from v9 as @pmario stated in #5961

I thought, there has been a PR somewhere. ... But I'm not sure atm

**Comment by nilslindemann** [comment]
I just tested with a [download containing all languages](https://highlightjs.org/download/). The file size of highlight.pack.js is then 1MB.

Aside of that, the highlighting works with my two changes in the highlightblock.js, mentioned above (these changes adapt to the breaking changes in the highlight.js project). Tested on tiddlywiki.com.

**Reply by pmario** [reply]
>... Aside of that, the highlighting works with my two changes in the highlightblock.js,

I don't know if highlight.js has code sanitation. I would like to avoid `.innerHTML()` if possible.

**Reply by nilslindemann** [reply]
@pmario I am not an expert in this.

I know that

* TiddlyWiki strips away script tags from WikiText
* it escapes `<` from start and end tags
* Setting `someElement.innerHTML` to `"<" + "script>alert('evil')</" + "script>";` [does not execute that script](https://stackoverflow.com/questions/1197575/can-scripts-be-inserted-with-innerhtml/).

While checking this, I found out that highlight.js [warns about unescaped HTML](https://github.com/highlightjs/highlight.js/blob/10b322dc436333b04df8295c9bdf4bad324849e0/src/highlight.js#L729-L747) in the console, when calling `hljs.highlightElement()` on elements which have child elements.

<strike>

I got rid of this warning by doing

```javascript
domNode.textContent = domNode.textContent;
hljs.highlightElement(domNode);
```

I do not know if this really solves the problem. [I have asked here](https://github.com/highlightjs/highlight.js/issues/2886#issuecomment-975931395). I do not know why highlight.js doesn't do this internally.

</strike>

**Edit**: that is actually not necessary, [as highlight.js does this already](https://github.com/highlightjs/highlight.js/blob/10b322dc436333b04df8295c9bdf4bad324849e0/src/highlight.js#L750). Instead – because TiddlyWiki properly escapes – we can replace [this line](https://github.com/Jermolene/TiddlyWiki5/blob/122de63b71a2454314365ac396e051ad1db77641/plugins/tiddlywiki/highlight/highlightblock.js#L21) with `hljs.configure({tabReplace: "    ", ignoreUnescapedHTML: true});`, getting rid of the warning.

See [here](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent#differences_from_innertext).

**Reply by nilslindemann** [reply]
This [SO Answer](https://stackoverflow.com/a/34064434/1658543) is also interesting. Instead of passing a string directly, it first parses that string with a [DOMParser](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) and then returns the `textContent` from that parsed document. Why this prevents XSS, I do not know. I guess a DOM `textNode` gets interpreted differently by `innerHTML` than a vanialla string.

**Comment by pmario** [comment]
There is a list of languages in the plugin readme, with those we usually include

**Comment by Jermolene** [comment]
Thanks @kookma 

> why Evernote is included? and why not highlight.js is include? Is there any reason?

I think it goes back to the very early days when we didn't have a plugin library, and so I would have to make sure that there was an edition containing each plugin so that users could drag and drop them. At that point, when I'd added a new plugin I'd generally leave it on tiddlywiki.com for a while for people to be able to test it.

It's probably time to standardise the plugins in /index.html and /prerelease.html (the latter is preloaded with many plugins).

**Reply by kookma** [reply]
Many thanks for clarification @Jermolene. Yes, it is time to standardise the plugins!
I would also much appreciate to have a look at https://github.com/Jermolene/TiddlyWiki5/discussions/6167 which is somehow related to distribution of official plugins and the way they appear in Sidebar and Control panel.
